### PR TITLE
Add package to hold Bucklescript bindings for ReasonML

### PR DESCRIPTION
**Pull request by emmenko**
> This PR supersedes #96 ([read the reasons behind it](https://github.com/tdeekens/flopflip/pull/96#issuecomment-355568791)).

The PR contains:
* new package setup (for bucklescript)
* the package name `@flopflip/reason-bindings` is still a **proposal**, it's not 100% decided yet
* a `example` folder that is meant to contain usage examples for each binding
* a module `FlopFlip__Types.re` that shares the types across the different components 
* bindings for `ConfigureFlopFlip` (react-broadcast)
  * a module `FlopFlip_ReactBroadcast__ConfigureFlopFlip.re` that directly exposes the react component (file naming is important!)
* (_bindings_) for `ToggleFeature` (react-broadcast)
  * a module `FlopFlip_ReactBroadcast__ToggleFeature.re` that implements a bit different API than its JS counterpart

> Read about [namespacing](https://github.com/tdeekens/flopflip/pull/106#issuecomment-355862017).
 
Other bindings will be provided in separate PRs

---
NOTE: we don't provide 1:1 bindings with each one of the components from each package. We only provide bindings **needed to work with the library** in Reason-land.

### CODE REVIEW

**Comment by tdeekens** [comment]
Nice, I like that we just offer bindings for parts of flopflip which make sense in Reason. All looks good to me. Thanks! I guess picking up the plan of the previous PR we just need to port `ToggleFeature`. I guess it's worth checking out `bs-react-router` as it's essentially the same API patterns.

**Comment by emmenko** [comment]
The RR bindings are actually very basics, as they are just binding components (I know because I’m also contributing to it 😉).

For our case, if we go with the approach I proposed, is a bit more complicated as it’s not a 1:1 binding. We can maybe discuss this in the next days in person, just to have a better understanding 🤓

Another question: are there any other components that we need to port? If so should we make a list of them?

**Comment by tdeekens** [comment]
> are there any other components that we need to port?

That depends on the package. For `react-broadcast` we need

- [x] `ConfigureFlopflip`
- [ ] `ToggleFeature`
- [ ] `SwitchFeature` (the same as `Switch` in react-router)

that's all a user of the lib needs. Setup and config using `ConfigureFlopflip` and toggling using `ToggleFeature`. Some other intermediary components would be used behind the scenes such as `FlagSubscription` to integrate with `react-broadcast` but I guess we don't need to bind then?

for `react-redux` the minimal set would be the same with some additional parts (depending on if we need to bind them).

- [x] `ConfigureFlopflip`
- [ ] `ToggleFeature`
- [ ] `SwitchFeature`
- [ ] `flopflipReducer` (to integrate into the store)
- [ ] `FLOPFLIP_STATE_SLICE ` (constant key of the reducer)
- [ ] `flopflipReducer` (to integrate into the store)
- [ ] `createFlopFlipEnhancer` (for people who don't want to configure using `ConfigureFlopflip`)

I would assume we would have to bind most setup things for redux. Obviously not all in one PR if that makes things easier.

**Comment by emmenko** [comment]
Understood!

For the redux part I would actually only bind the components.
The “reducer” parts I guess are used to configure redux. So unless you are also using redux in reason (which I haven’t seen a library yet) you will still keep that part in JS, hence there is no need to bind those things yet and I would suggest to get back to those once there is a need for it.

WDYT?

**Comment by tdeekens** [comment]
Sure makes total sense. I did strike them from the above list

**Comment by emmenko** [comment]
Short update (see last commit https://github.com/tdeekens/flopflip/pull/106/commits/7007a4b9a2b0ab689edba20a1f4945b528154769): we need to introduce namespacing, especially because within the library itself there are different exports of the same component (e.g. `ConfigureFlopFlip`) from different packages.

So at the end you would use the modules like:

```reason
<FlopFlip_ReactBroadcast.ConfigureFlopFlip />

<FlopFlip_ReactRedux.ConfigureFlopFlip />
```

**Comment by tdeekens** [comment]
Sorry for not being on top of this. The namespacing manual namespacing makes total sense until the scoped package namespacing is improved.

File: packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  let make = (~channel: string, children) =>
+    ReasonReact.wrapJsForReason(
+      ~reactClass,
+      ~props={"channel": channel},
+      children
+    );
+};
+
+[@bs.module "@flopflip/react"]
+external isFeaturedEnabled :
```
Comment (tdeekens) [inline packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re]: IsFeatureEnabled.

File: packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re [hunk 2]
```diff
@@ -0,0 +1,40 @@
+open FlopFlip__Types;
+
+module Subscriber = {
+  [@bs.module "react-broadcast"]
+  external reactClass : ReasonReact.reactClass = "Subscriber";
```
Comment (tdeekens) [inline packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re]: Do we have to bind this component as it‘s used by ToggleFeature?

File: packages/reason-bindings/examples/example_configure_flopflip.re [hunk 3]
```diff
@@ -1,4 +1,4 @@
-open FlopFlipTypes;
+open FlopFlip__Types;
```
Comment (tdeekens) [inline packages/reason-bindings/examples/example_configure_flopflip.re]: Silly question: is the double underscore intentional?

File: packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re (hunk 2 above)
Comment (emmenko) [reply packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re]: Well yes, the ToggleFeature implementation here is different from the original one.

File: packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re (hunk 1 above)
Comment (emmenko) [reply packages/reason-bindings/src/FlopFlip_ReactBroadcast__ToggleFeature.re]: Ups

File: packages/reason-bindings/examples/example_configure_flopflip.re (hunk 3 above)
Comment (emmenko) [reply packages/reason-bindings/examples/example_configure_flopflip.re]: Yes, apparently is some kind of convention for namespacing

**Comment by tdeekens** [comment]
I am closing this. I am very thankful for the contribution but I lost track of the progress and status nor has anybody else demanded this. I would love to re-open this once work on this is active again.

**Pull request closed without merging**